src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -867,28 +921,33 @@ private void WriteHeaders(HttpRequestMessage request)
             }
         }
 
-        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request)
+        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             // Ensure we don't exceed the max concurrent streams setting.
-            await _concurrentStreams.RequestCreditAsync(1).ConfigureAwait(false);
+            await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
 
-            // Note, HEADERS and CONTINUATION frames must be together, so hold the writer lock across sending all of them.
-            // We also serialize usage of the header encoder and the header buffer this way.
-            // (If necessary, we could have a separate semaphore just for creating and encoding header blocks,
-            // and defer taking the actual _writerLock until we're ready to do the write below.)
-            await _writerLock.WaitAsync().ConfigureAwait(false);
+            // We serialize usage of the header encoder and the header buffer seperately from the
+            // write lock
+            await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
It's hard for me to tell... do we have a strict ordering as to the acquisition of _headerSerializationLock and _writerLock?  We need to make sure that if one can be held while acquiring the other that we always do so in the same order; otherwise we risk deadlock.

Reply:
Also, can you clarify why a separate lock is needed?  Why can't it just hold the single write lock across both operations?

Reply:
There is a strict ordering to the locks here. _headerSerializationLock is only needed for this function, and will always be acquired before _writerLock.

The goal of the _headerSerializationLock is to protect the header buffer, and other associated resources. We can only have a single thread accessing those resources at a time, and it is not safe to release that lock until that buffer has been flushed to the write buffer, which is why we end up acquiring both locks here.

On the other hand, it is actually safe for threads to write other (non header) data to the write buffer while we're serializing headers. As long as they don't touch the header, that's a-ok. Having a seperate lock allows us to do that.